config.go

```diff
@@ -27,19 +27,88 @@ type config struct {
 	DisableCheckpointSignature bool `json:"disable_checkpoint_signature"`
 	PluginMinPort              int
 	PluginMaxPort              int
-
-	Builders       packer.MapOfBuilder
-	Provisioners   packer.MapOfProvisioner
-	PostProcessors packer.MapOfPostProcessor `json:"post-processors"`
+	RawBuilders                map[string]string         `json:"builders"`
+	RawProvisioners            map[string]string         `json:"provisioners"`
+	RawPostProcessors          map[string]string         `json:"post-processors"`
+	Builders                   packer.MapOfBuilder       `json:"-"`
+	Provisioners               packer.MapOfProvisioner   `json:"-"`
+	PostProcessors             packer.MapOfPostProcessor `json:"-"`
 }
 
-// Decodes configuration in JSON format from the given io.Reader into
+// decodeConfig decodes configuration in JSON format from the given io.Reader into
 // the config object pointed to.
 func decodeConfig(r io.Reader, c *config) error {
 	decoder := json.NewDecoder(r)
 	return decoder.Decode(c)
 }
 
+// LoadExternalComponentsFromConfig loads plugins defined in RawBuilders, RawProvisioners, and RawPostProcessors.
+func (c *config) LoadExternalComponentsFromConfig() {
+	// helper to build up list of plugin paths
+	extractPaths := func(m map[string]string) []string {
+		paths := make([]string, 0, len(m))
+		for _, v := range m {
+			paths = append(paths, v)
+		}
+
+		return paths
+	}
+
+	var pluginPaths []string
+	pluginPaths = append(pluginPaths, extractPaths(c.RawProvisioners)...)
+	pluginPaths = append(pluginPaths, extractPaths(c.RawBuilders)...)
+	pluginPaths = append(pluginPaths, extractPaths(c.RawPostProcessors)...)
+
+	var externallyUsed = make([]string, 0, len(pluginPaths))
+	for _, pluginPath := range pluginPaths {
+		if name, ok := c.loadExternalComponent(pluginPath); ok {
```

Comment:
I originally used `discoverExternalComponents(filepath.Dir(pluginPath))` here to load the plugins, but found that this introduced an unexpected behavior of loading any plugin within the parent directory that might not be defined within the packer config. The `loadExternalComponent` (naming is hard) function essentially just loads one plugin. TBH the code is very similar  to the contents of discoverSingle and discoverExternalCompenets so there might be a simpler path or an opportunity to consolidate.